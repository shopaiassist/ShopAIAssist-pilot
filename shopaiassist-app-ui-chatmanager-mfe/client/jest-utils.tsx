import React, { JSXElementConstructor, ReactElement } from 'react';
import { Provider } from 'react-redux';
import { MemoryRouterProps } from 'react-router';
import { MemoryRouter } from 'react-router-dom';
import { I18nextProvider } from 'react-i18next';
import { merge } from 'ts-deepmerge';
import { configureStore, EnhancedStore } from '@reduxjs/toolkit';
import { ByRoleMatcher, render, RenderResult } from '@testing-library/react';
import { reducer, RootState, store as baseStore } from './src/store/store';
import i18nInstance from './src/utils/i18n';

type InitialEntries = MemoryRouterProps['initialEntries'];
type DeepPartial<T> = {
  [P in keyof T]?: T[P] extends object ? DeepPartial<T[P]> : T[P];
};
type PartialRootState = DeepPartial<RootState>;

interface RenderOptions {
  entries?: InitialEntries;
  state?: PartialRootState;
  store?: EnhancedStore;
}

interface SnapshotOptions {
  isModal?: boolean | ByRoleMatcher;
  state?: PartialRootState;
}

interface WrapperProps {
  children: ReactElement;
}

// export const mockApi = (apiInstance: AxiosInstance) => ({
//   delete: jest.spyOn(apiInstance, 'delete'),
//   get: jest.spyOn(apiInstance, 'get'),
//   post: jest.spyOn(apiInstance, 'post'),
//   put: jest.spyOn(apiInstance, 'put'),
//   patch: jest.spyOn(apiInstance, 'patch'),
// });

/** Redux store with optional default content */
export const mockStore = (state: PartialRootState = {}): EnhancedStore =>
  configureStore({ preloadedState: merge(baseStore.getState(), state), reducer: reducer as never });

/** Render function with access to override store/state */
export const renderAll = (ui: ReactElement, { entries, state = {}, store }: RenderOptions = {}): RenderResult => {
  const data = store || mockStore(state);
  const Wrapper = ({ children }: WrapperProps) => (
    <MemoryRouter initialEntries={entries}>
      <Provider store={data}>
        <I18nextProvider i18n={i18nInstance}>{children}</I18nextProvider>
      </Provider>
    </MemoryRouter>
  );
  return render(ui, { wrapper: Wrapper as JSXElementConstructor<{ children: React.ReactNode }> });
};

/** Render function with i18n provider */
export const renderWithI18n = (ui: ReactElement): RenderResult => {
  const I18nWrapper = ({ children }: WrapperProps) => <I18nextProvider i18n={i18nInstance}>{children}</I18nextProvider>;
  return render(ui, { wrapper: I18nWrapper as JSXElementConstructor<{ children: React.ReactNode }> });
};

/** Render function with access to override store/state */
export const renderWithRedux = (ui: ReactElement, { state = {}, store }: RenderOptions = {}): RenderResult => {
  const data = store || mockStore(state);
  const ReduxWrapper = ({ children }: WrapperProps) => (
    <Provider store={data}>
      <I18nextProvider i18n={i18nInstance}>{children}</I18nextProvider>
    </Provider>
  );
  return render(ui, { wrapper: ReduxWrapper as JSXElementConstructor<{ children: React.ReactNode }> });
};

/** Get content to test against a snapshot */
export const getSnapshot = (ui: ReactElement, state?: PartialRootState): ChildNode =>
  renderAll(ui, { state }).container.firstChild as ChildNode;

/** Get modal content to test against a snapshot */
export const getSnapshotModal = (ui: ReactElement, state?: PartialRootState, role?: ByRoleMatcher): ChildNode =>
  renderAll(ui, { state }).getByRole(role || 'presentation');

/** Helper function to generate snapshots */
export const snapshot = (ui: ReactElement, { isModal, state }: SnapshotOptions = {}) => {
  const role = typeof isModal === 'boolean' ? undefined : isModal;
  // eslint-disable-next-line no-extra-boolean-cast
  const snapshot = !!isModal ? getSnapshotModal(ui, state, role) : getSnapshot(ui, state);
  expect(snapshot).toMatchSnapshot();
};

// re-export everything
export * from '@testing-library/react';
