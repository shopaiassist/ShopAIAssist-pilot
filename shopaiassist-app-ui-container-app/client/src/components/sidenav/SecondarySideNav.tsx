import React from 'react';
import { i18n, Translate } from '@/';
import { SafButton, SafIcon, SafSideNav, SafText } from '@/core-components/react';
import { useNavigate } from 'react-router-dom';

import { OLYMPUS_ROUTES } from '../../App';
import useSelectedMenuItem from '../../hooks/useSelectedMenuItem';
import { getPathName } from '../../utils/misc';
import NavMenuItem from './NavMenuItem';

import './SideNav.scss';

interface SecondarySideNavProps {
  tree: SecondarySideNavTree;
  isMobile: boolean;
  closeNav: () => void;
  toggleMainNav: (open?: boolean) => void;
}

const SecondarySideNav = ({ tree, isMobile, closeNav, toggleMainNav }: SecondarySideNavProps) => {
  const navigate = useNavigate();
  const { t } = i18n.useTranslation();

  /** Support info config */
  const SupportInfo = {
    SUPPORT_NUMBER: '8332273898',
    SUPPORT_EMAIL: 'support@.com',
    TRAINING_LINK: 'https://training.example.com/legal--core',
    HELP_ARTICLES_LINK: 'https://www.example.com/en-us/help/.html',
    SUPPORT_HOURS: {
      weekDays: {
        starting: 7,
        ending: 22,
        timezone: 'ET',
      },
      weekend: {
        starting: 8,
        ending: 17,
        timezone: 'ET',
      },
    },
  };

  /** Formats a number representing a 24hrs format hour to a string in 12hrs format */
  const formatHour = (hour: number) => {
    if (hour < 12) return `${hour} AM`;
    return `${hour - 12} PM`;
  };

  /**
   * Formats a plain string representing a 10-digit phone number.
   * Example:
   *   1234567890 -> (123) 456-7890
   */
  const formatPhoneNumber = (number: string) => {
    const cleaned = number.replace(/\\D/g, '');
    const match = cleaned.match(/^(\d{3})(\d{3})(\d{4})$/);
    if (match) {
      number = '(' + match[1] + ') ' + match[2] + '-' + match[3];
    }
    return `${t('CALL')} ${number}`;
  };

  /** Toggle the OneTrust Cookie Settings display  */
  const onCookieSettingsClick = () => {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    (global as any).OneTrust?.ToggleInfoDisplay();
  };

  /** List of nav menu items to render */
  const navMenuItems: NavMenuTree = {
    manage: [
      {
        id: 'favorites-nav-item',
        label: t('FAVORITES'),
        navPath: OLYMPUS_ROUTES.MANAGE_FAVORITES,
      },
    ],
    admin: [
      {
        id: 'permissions-nav-item',
        label: t('USER_PERMISSIONS'),
        navPath: OLYMPUS_ROUTES.ADMIN_PERMISSIONS,
      },
      {
        id: 'analytics-nav-item',
        label: t('USAGE_ANALYTICS'),
        navPath: OLYMPUS_ROUTES.ADMIN_USAGE,
      },
      {
        id: 'integrations-nav-item',
        label: t('INTEGRATIONS'),
        navPath: OLYMPUS_ROUTES.ADMIN_INTEGRATIONS,
      },
    ],
    support: [
      {
        id: 'live-chat-nav-item',
        icon: 'comments',
        label: t('LIVE_CHAT'),
      },
      {
        id: 'phone-nav-item',
        label: formatPhoneNumber(SupportInfo.SUPPORT_NUMBER),
        url: `tel:${SupportInfo.SUPPORT_NUMBER}`,
      },
      {
        id: 'email-nav-item',
        label: t('EMAIL'),
        url: `mailto:${SupportInfo.SUPPORT_EMAIL}`,
      },
      {
        id: 'help-articles-nav-item',
        label: t('HELP_ARTICLES'),
        url: SupportInfo.HELP_ARTICLES_LINK,
      },
      {
        id: 'training-nav-item',
        label: t('TRAINING'),
        url: SupportInfo.TRAINING_LINK,
      },
    ],
    legal: [
      {
        id: 'terms-nav-item',
        label: t('TERMS'),
        url: 'https://www.example.com/en/terms-of-use.html',
      },
      {
        id: 'privacy-nav-item',
        label: t('PRIVACY'),
        url: 'https://www.example.com/en/privacy-statement.html',
      },
      {
        id: 'cookie-settings-nav-item',
        label: t('COOKIE_SETTINGS'),
        onClick: onCookieSettingsClick,
      },
      {
        id: 'cookie-policy-nav-item',
        label: t('COOKIE_POLICY'),
        url: 'https://www.example.com/en/privacy-statement.html#CookieIBA',
      },
      {
        id: 'do-not-sell-nav-item',
        label: t('DO_NOT_SELL'),
        url: '/redirect/ccpa-dsar',
      },
    ],
  };

  const [selectedMenuItem, setSelectedMenuItem] = useSelectedMenuItem(navMenuItems);

  /** Handle selecting menu items */
  const handleMenuItemClick = (menuItem: NavMenuItem) => {
    setSelectedMenuItem(menuItem.id);
    if (menuItem.navPath) {
      if (isMobile) {
        closeNav();
        toggleMainNav(false);
      }
      navigate(menuItem.navPath);
    }
    menuItem.onClick?.();
  };

  /** Function to render a NavMenuItemComponent */
  const renderNavMenuItem = (menuItem: NavMenuItem) => {
    const { bold, id, icon, label, showArrow, url } = menuItem;
    return (
      <NavMenuItem
        id={id}
        key={id}
        onClick={() => handleMenuItemClick(menuItem)}
        icon={icon}
        label={label}
        showArrow={showArrow}
        selected={selectedMenuItem === id}
        url={url}
        useTooltip={false}
        bold={bold}
      />
    );
  };

  return (
    <SafSideNav
      id="secondary-side-nav"
      className="olympus-side-nav secondary-side-nav expanded"
      state={'open'}
      aria-label="Secondary navigation"
    >
      <header className="header">
        <SafText appearance="body-strong-md" style={{ paddingLeft: '0.75rem' }}>
          {getPathName(tree, t)}
        </SafText>
        <SafButton iconOnly appearance="tertiary" onClick={closeNav} aria-label="Close secondary navigation">
          <SafIcon icon-name="close" size={16} />
        </SafButton>
      </header>
      <div className="secondary-nav-container">
        {tree === 'support' && (
          <div className="nav-group">
            <div className="support-info">
              <SafText appearance="body-strong-sm">
                <Translate tKey="SUPPORT_HOURS" />
              </SafText>
              <div className="help-menu-item">
                <SafText appearance="body-default-sm">
                  {`Mon-Fri ${formatHour(SupportInfo.SUPPORT_HOURS.weekDays.starting)} - ${formatHour(SupportInfo.SUPPORT_HOURS.weekDays.ending)} ${SupportInfo.SUPPORT_HOURS.weekDays.timezone}`}
                </SafText>
                <br />
                <SafText appearance="body-default-sm">
                  {`Sat-Sun ${formatHour(SupportInfo.SUPPORT_HOURS.weekend.starting)} - ${formatHour(SupportInfo.SUPPORT_HOURS.weekend.ending)} ${SupportInfo.SUPPORT_HOURS.weekend.timezone}`}
                </SafText>
              </div>
            </div>
          </div>
        )}
        <div className="nav-group">{navMenuItems[tree].map(renderNavMenuItem)}</div>
      </div>
    </SafSideNav>
  );
};

export default SecondarySideNav;
