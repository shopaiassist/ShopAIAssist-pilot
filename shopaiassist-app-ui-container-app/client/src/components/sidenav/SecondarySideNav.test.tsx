// eslint-disable-next-line @typescript-eslint/no-unused-vars
import { renderAll, snapshot, waitFor } from '../../../jest-utils';
import SecondarySideNav from './SecondarySideNav';

describe('SecondarySideNav', () => {
  const mockOnClose = jest.fn();

  describe('render', () => {
    it('should render manage tree', async () => {
      snapshot(<SecondarySideNav tree="manage" closeNav={mockOnClose} isMobile={false} toggleMainNav={() => {}} />);
    });
    it('should render admin tree', async () => {
      snapshot(<SecondarySideNav tree="admin" closeNav={mockOnClose} isMobile={false} toggleMainNav={() => {}} />);
    });
    it('should render support tree', async () => {
      snapshot(<SecondarySideNav tree="support" closeNav={mockOnClose} isMobile={false} toggleMainNav={() => {}} />);
    });
    it('should render legal tree', async () => {
      snapshot(<SecondarySideNav tree="legal" closeNav={mockOnClose} isMobile={false} toggleMainNav={() => {}} />);
    });
  });
});
