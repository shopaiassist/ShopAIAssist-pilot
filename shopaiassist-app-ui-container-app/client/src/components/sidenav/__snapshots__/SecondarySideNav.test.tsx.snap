// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SecondarySideNav render should render admin tree 1`] = `
<div
  aria-label="Secondary navigation"
  class="olympus-side-nav secondary-side-nav expanded"
  id="secondary-side-nav"
  state="open"
>
  <header
    class="header"
  >
    <span
      appearance="body-strong-md"
      style="padding-left: 0.75rem;"
    >
      Admin settings
    </span>
    <button
      appearance="tertiary"
      aria-label="Close secondary navigation"
    >
      <div
        icon-name="close"
        size="16"
      />
    </button>
  </header>
  <div
    class="secondary-nav-container"
  >
    <div
      class="nav-group"
    >
      <div
        class="nav-menu-item"
        id="permissions-nav-item"
        target="_blank"
      >
        <div
          class="nav-item-content"
        >
          <span
            appearance="body-default-md"
          >
            User permissions
          </span>
        </div>
      </div>
      <div
        class="nav-menu-item"
        id="analytics-nav-item"
        target="_blank"
      >
        <div
          class="nav-item-content"
        >
          <span
            appearance="body-default-md"
          >
            Usage analytics
          </span>
        </div>
      </div>
      <div
        class="nav-menu-item"
        id="integrations-nav-item"
        target="_blank"
      >
        <div
          class="nav-item-content"
        >
          <span
            appearance="body-default-md"
          >
            Integrations
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`SecondarySideNav render should render legal tree 1`] = `
<div
  aria-label="Secondary navigation"
  class="olympus-side-nav secondary-side-nav expanded"
  id="secondary-side-nav"
  state="open"
>
  <header
    class="header"
  >
    <span
      appearance="body-strong-md"
      style="padding-left: 0.75rem;"
    >
      Legal information
    </span>
    <button
      appearance="tertiary"
      aria-label="Close secondary navigation"
    >
      <div
        icon-name="close"
        size="16"
      />
    </button>
  </header>
  <div
    class="secondary-nav-container"
  >
    <div
      class="nav-group"
    >
      <div
        class="nav-menu-item"
        id="terms-nav-item"
        target="_blank"
        url="https://www.example.com/en/terms-of-use.html"
      >
        <div
          class="nav-item-content"
        >
          <span
            appearance="body-default-md"
          >
            Terms of use
          </span>
          <div
            style="position: absolute; right: 0.5rem;"
          >
            <div
              appearance="light"
              icon-name="external-link"
              size="16"
            />
          </div>
        </div>
      </div>
      <div
        class="nav-menu-item"
        id="privacy-nav-item"
        target="_blank"
        url="https://www.example.com/en/privacy-statement.html"
      >
        <div
          class="nav-item-content"
        >
          <span
            appearance="body-default-md"
          >
            Privacy
          </span>
          <div
            style="position: absolute; right: 0.5rem;"
          >
            <div
              appearance="light"
              icon-name="external-link"
              size="16"
            />
          </div>
        </div>
      </div>
      <div
        class="nav-menu-item"
        id="cookie-settings-nav-item"
        target="_blank"
      >
        <div
          class="nav-item-content"
        >
          <span
            appearance="body-default-md"
          >
            Cookie Settings
          </span>
        </div>
      </div>
      <div
        class="nav-menu-item"
        id="cookie-policy-nav-item"
        target="_blank"
        url="https://www.example.com/en/privacy-statement.html#CookieIBA"
      >
        <div
          class="nav-item-content"
        >
          <span
            appearance="body-default-md"
          >
            Cookie Policy
          </span>
          <div
            style="position: absolute; right: 0.5rem;"
          >
            <div
              appearance="light"
              icon-name="external-link"
              size="16"
            />
          </div>
        </div>
      </div>
      <div
        class="nav-menu-item"
        id="do-not-sell-nav-item"
        target="_blank"
        url="/redirect/ccpa-dsar"
      >
        <div
          class="nav-item-content"
        >
          <span
            appearance="body-default-md"
          >
            Do not sell or share my personal information and limit the use of my sensitive personal information
          </span>
          <div
            style="position: absolute; right: 0.5rem;"
          >
            <div
              appearance="light"
              icon-name="external-link"
              size="16"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`SecondarySideNav render should render manage tree 1`] = `
<div
  aria-label="Secondary navigation"
  class="olympus-side-nav secondary-side-nav expanded"
  id="secondary-side-nav"
  state="open"
>
  <header
    class="header"
  >
    <span
      appearance="body-strong-md"
      style="padding-left: 0.75rem;"
    >
      Manage ShopAIAssist
    </span>
    <button
      appearance="tertiary"
      aria-label="Close secondary navigation"
    >
      <div
        icon-name="close"
        size="16"
      />
    </button>
  </header>
  <div
    class="secondary-nav-container"
  >
    <div
      class="nav-group"
    >
      <div
        class="nav-menu-item"
        id="favorites-nav-item"
        target="_blank"
      >
        <div
          class="nav-item-content"
        >
          <span
            appearance="body-default-md"
          >
            Favorites
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`SecondarySideNav render should render support tree 1`] = `
<div
  aria-label="Secondary navigation"
  class="olympus-side-nav secondary-side-nav expanded"
  id="secondary-side-nav"
  state="open"
>
  <header
    class="header"
  >
    <span
      appearance="body-strong-md"
      style="padding-left: 0.75rem;"
    >
      Support
    </span>
    <button
      appearance="tertiary"
      aria-label="Close secondary navigation"
    >
      <div
        icon-name="close"
        size="16"
      />
    </button>
  </header>
  <div
    class="secondary-nav-container"
  >
    <div
      class="nav-group"
    >
      <div
        class="support-info"
      >
        <span
          appearance="body-strong-sm"
        >
          Customer support hours
        </span>
        <div
          class="help-menu-item"
        >
          <span
            appearance="body-default-sm"
          >
            Mon-Fri 7 AM - 10 PM ET
          </span>
          <br />
          <span
            appearance="body-default-sm"
          >
            Sat-Sun 8 AM - 5 PM ET
          </span>
        </div>
      </div>
    </div>
    <div
      class="nav-group"
    >
      <div
        class="nav-menu-item"
        id="live-chat-nav-item"
        target="_blank"
      >
        <div
          appearance="light"
          icon-name="comments"
          size="16"
          slot="start"
          style="align-items: center;"
        />
        <div
          class="nav-item-content"
        >
          <span
            appearance="body-default-md"
          >
            Live chat support
          </span>
        </div>
      </div>
      <div
        class="nav-menu-item"
        id="phone-nav-item"
        target="_blank"
        url="tel:8332273898"
      >
        <div
          class="nav-item-content"
        >
          <span
            appearance="body-default-md"
          >
            Call (833) 227-3898
          </span>
          <div
            style="position: absolute; right: 0.5rem;"
          >
            <div
              appearance="light"
              icon-name="external-link"
              size="16"
            />
          </div>
        </div>
      </div>
      <div
        class="nav-menu-item"
        id="email-nav-item"
        target="_blank"
        url="mailto:support@.com"
      >
        <div
          class="nav-item-content"
        >
          <span
            appearance="body-default-md"
          >
            Email
          </span>
          <div
            style="position: absolute; right: 0.5rem;"
          >
            <div
              appearance="light"
              icon-name="external-link"
              size="16"
            />
          </div>
        </div>
      </div>
      <div
        class="nav-menu-item"
        id="help-articles-nav-item"
        target="_blank"
        url="https://www.example.com/en-us/help/.html"
      >
        <div
          class="nav-item-content"
        >
          <span
            appearance="body-default-md"
          >
            Help articles
          </span>
          <div
            style="position: absolute; right: 0.5rem;"
          >
            <div
              appearance="light"
              icon-name="external-link"
              size="16"
            />
          </div>
        </div>
      </div>
      <div
        class="nav-menu-item"
        id="training-nav-item"
        target="_blank"
        url="https://training.example.com/legal--core"
      >
        <div
          class="nav-item-content"
        >
          <span
            appearance="body-default-md"
          >
            Training
          </span>
          <div
            style="position: absolute; right: 0.5rem;"
          >
            <div
              appearance="light"
              icon-name="external-link"
              size="16"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
